### DISCUSSION TITLE
How RasterDataset() for image folder and mask folder matches corresponding image and mask pair

### DISCUSSION THREAD

**Discussion by Raji-Ahmad** [torchgeo/torchgeo#1014] [Q&A] [answered]
I tried to understand how RasterDataset class indexes files in a folder to be sure if when called on image folder and mask folder will index (match) them correctly in dataLoader. For example;

ds = RasterDataset(image_f) & RasterDataset(mask_f)
sampler = GridGeoSampler(ds, size=patch_size, stride=stride)
trainLoader = DataLoader(ds, sampler=sampler)

How is the indexing done? Is it order of filename such that both image and mask should have same filenames or by spatial bounds?

**Comment by adamjstewart** [comment, marked as answer]
You might want to read https://arxiv.org/abs/2111.08872

TL;DR: all GeoDatasets (including RasterDataset and VectorDataset) read the geographic metadata from each file and build a spatiotemporal R-tree. The sampler gets this R-tree from the dataset and decides where to sample from. The sampler chooses a spatiotemporal bounding box, and each dataset checks its own R-tree to see which files overlap with that bounding box. Then we use rasterio windowed reading to automatically merge all files at that location into a single image and/or mask.

**Reply by Raji-Ahmad** [reply]
I understood what you explained. But in a scenario where I have two images of different non-adjacent areas (with existing gaps between them), is it possible for the sampler to sample a bounding box that intersects just a little part of one image and the rest part of the box being part of the empty gaps?

I want to know how some things work and not leave my app on some guesswork.

**Reply by adamjstewart** [reply]
GeoSamplers only sample from regions where there is data, they won't sample from empty gaps between images. The R-tree will contain an entry for the bounding box of every file in the dataset. Depending on which sampler you use, they will choose a random/sequential file from the dataset and sample patches randomly/sequentially within the area of that file.

**Reply by Raji-Ahmad** [reply]
That answers my question. Thanks.
